boost: add host library build for boost

ammubhave opened the pull request:
Summary: Currently, boost host build doesn't actually compile boost
libraries for the host arch. If there are host tools that would want
to link against boost, it can't right now.

I am currently proposing https://github.com/openwrt/packages/pull/8949 which requires the boost_{filesystem,system,thread} libraries compiled for the host target, since that PR is for a host tool which is a thrift compiler.

This diff adds support for compiling host boost library as well.
Any packages that need to depend on a host boost library can depend on
CONFIG_boost-host-build-<lib>.

Test Plan: Compiled https://github.com/openwrt/packages/pull/8949 and verified that it works.
used openwrt master using nbg6817

Maintainer: @ClaymorePT

Signed-off-by: Amol Bhave <ambhave@fb.com>

ClaymorePT commented:
Ok, I have some issues and questions about this PR.

The boost package is supposed to be a package for a specific target embedded system.
This means all the package selected  libraries will affect the package built for the target system.
What your PR does is building boost for the host system with the options which were meant for the target system.

If you have a host tool which requires Boost.Fiber for example, this means Boost.Fiber and its dependencies will also be built for the target system, even if they are not required. This behaviour should not happen.

Also, if it's a host tool that you want to build linked to boost, why not just use the host boost library in the first place?

@neheb Can you give some hints here? Should OpenWRT be building Boost for host tools? Why not just use the host boost library.

ClaymorePT commented:
@ammubhave Ok, this PR is interesting.

@neheb Can you give a hint here? Is OpenWRT currently building its host tools linked with host libraries?

ammubhave commented:
Updated description and added comments for clarity. I am proposing https://github.com/openwrt/packages/pull/8949/ which is host tool, a thrift compiler. This requires boost_{filesystem,system,thread} libraries compiled for the host target.

The boost headers are installed for the host unconditionally because I noticed that packages such as https://github.com/openwrt/packages/blob/master/net/kea/Makefile already exist which are host tools and depend on boost headers. However, currently they get by with a hack `--with-boost-include="$(STAGING_DIR)/usr/include"`. In favor not modifying such packages to explicitly state to install boost headers for the host, I defaulted to always including the headers.

neheb commented:
The PR looks sound. I don't really use or develop with Boost so I can't comment on specifics.

ClaymorePT commented:
@neheb I was asking more regarding the way how host tools are built for OpenWRT.

In my opinion, you can merge this PR.
